Update to tensorflow.js

cvalenzuela opened the pull request:
Not ready to merge yet!

Following the update to the library (https://github.com/ml5js/ml5-library/pull/117), this PR updates the examples to match the new changes.

Major things are:
- KNN replaced for Transfer Learning: including training and loss elements
- Adding YOLO

I'm using a `/temp_lib` folder to point to the WIP `ml5.js 0.0.4` for now

shiffman commented on p5js/ImageClassification/sketch.js:
```diff
@@ -5,12 +5,11 @@
-// Initialize the ImageNet method with the MobileNet model.
-const classifier = new ml5.ImageClassifier('MobileNet');
+// Initialize the Image Classifier method
+const classifier = new ml5.ImageClassifier();
```
Is "MobileNet" the default but the Image classifier could still load other models as well? I think I like the idea of requiring a reference to the pre-trained model to increase transparency. And this also opens up the possibility of creating an "empty" Image Classifier that has to be trained from scratch?

cvalenzuela replied:
It is the default because Squeezenet does not currently work in tf.js (See: https://github.com/tensorflow/tfjs-core/tree/master/models)
I also liked specifying the model, we can bring it back!

shiffman commented on p5js/ImageClassification/sketch.js:
```diff
@@ -29,6 +28,6 @@ function imageReady() {
 function gotResult(results) {
   // The results are in an array ordered by probability.
-  select('#result').html(results[0].label);
+  select('#result').html(results[0].className);
```
Will "className" be more confusing than "label"? Is it important to use the underlying terminology? (I mean these to be open questions.)

cvalenzuela replied:
I change this because I came across much more references to 'class' and 'class name' rather than 'label' in other examples.

shiffman commented on p5js/ImageClassification_TransferLearning/sketch.js:
```diff
@@ -0,0 +1,94 @@
+
+// Add the current frame from the video to the classifier
+function addImage(label) {
+  classifier.addImage(label);
```
Note here we are using "label" as a variable name.

shiffman commented on p5js/ImageClassification_TransferLearning/sketch.js:
```diff
@@ -0,0 +1,94 @@
+function setup() {
+  noCanvas();
+  // Create a video element
+  video = createCapture(VIDEO, onVideoLoaded).parent('videoContainer');
```
Chaining can sometimes be a little intimidating to beginners.

cvalenzuela replied:
yep, agree! will update then.

shiffman commented on p5js/ImageClassification_TransferLearning/sketch.js:
```diff
@@ -0,0 +1,94 @@
+let catImages = 0;
+
+// Default options for training
+const options = {
```
I wonder if we rely on the defaults for our basic example and leave this out. Great to allow this possibility but a little scary looking. I guess now with transfer learning we have to specify the number of classes in advance?

cvalenzuela replied:
These are totally optional, no need to provide them (actually they are the same in the src code). But if you are using more classes and examples in your training you might need to tweak this values a bit.

The defaults expect 2 classes.

shiffman replied:
I wonder if there is some clever way we could wait to configure the number of new classes in the model until right before training.

This is on the borderline of perhaps too much magic behind the scenes, but having a flexible "teachable machine" where the interaction is making up classes and adding training images on the fly might be a useful feature.

cvalenzuela replied:
yes, totally. we could have predefined values that are updated based on the number of classes and images. Values we can get running a couple of examples and seeing performance.

shiffman commented on p5js/ImageClassification_TransferLearning/sketch.js:
```diff
@@ -0,0 +1,94 @@
+  // Add image as dog class
+  buttonB = select('#buttonB');
+  buttonB.mousePressed(function() {
+    addImage('dog');
```
Yay for string labels!!!

shiffman commented on p5js/ImageClassification_TransferLearning/sketch.js:
```diff
@@ -0,0 +1,94 @@
+
+  // Train Button
+  train = select('#train');
+  train.mousePressed(function() {
```
These nested callbacks are a little confusing, I can take a look at some point.

shiffman commented on p5js/YOLO/sketch.js:
```diff
@@ -0,0 +1,47 @@
+  createCanvas(640, 410);
+  video = createCapture(VIDEO, onVideoLoaded);
+  video.hide();
+  select('#start').mousePressed(function() {
```
Maybe just pass in detect directly? 

```javascript
select('#start').mousePressed(detect);
```

cvalenzuela replied:
ok, I'm still having some issues with the speed (FPS). Although the draw in not waiting for yolo to send results. We can see this tomorrow.

shiffman commented on p5js/YOLO/sketch.js:
```diff
@@ -0,0 +1,47 @@
+
+function draw() {
+  image(video, 0, 0, 640, 410);
+  objects.forEach(function(object) {
```
`forEach` loops aren't commonly used in p5 examples so I might just use a regular for loop or a `for..of` loop?

shiffman commented on p5js/YOLO/sketch.js:
```diff
@@ -0,0 +1,47 @@
+}
+
+function onVideoLoaded() {
+  yolo = new ml5.YOLO(video.elt);
```
A reminder to maybe revisit the question of if we can say things like `yolo = new ml5.YOLO(video)` for integration with p5?  Or is that opening too big of a can of worms? Sorry to be obsessed with this small detail :smile:

https://github.com/ml5js/ml5-library/issues/27

cvalenzuela replied:
Yes, totally! Just need to update the library.

cvalenzuela commented:
Just pushed new updates reflecting the changes to https://github.com/ml5js/ml5-library/pull/117

shiffman commented on p5js/LSTM_Interactive/index.html:
```diff
@@ -13,7 +13,8 @@
   <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/0.6.0/addons/p5.dom.min.js"></script>
-  <script src="https://unpkg.com/ml5@0.0.3/dist/ml5.js"></script>
+  <!-- <script src="https://unpkg.com/ml5@0.0.3/dist/ml5.js"></script> -->
+  <script src="../../temp_lib/ml5.min.js"></script>
```
This is b/c we haven't done the release yet?

cvalenzuela replied:
yep, just temporary

shiffman commented on p5js/PoseNet/sketch.js:
```diff
@@ -0,0 +1,67 @@
+
+// A function to draw ellipses over the detected keypoints
+function drawKeypoints() {
+  poses.forEach(pose => {
```
I've started using this style/syntax more now myself, but I think for the p5 examples  we should probably stick with more vanilla for loops or `for of` loops.

cvalenzuela replied:
Yes, agree. Will update this.

shiffman commented on p5js/PoseNet/sketch.js:
```diff
@@ -0,0 +1,67 @@
+// The callback that gets called every time there's an update from the model
+function gotPoses(results) {
+  poses = results;
+  skeletons = poses.map(pose => poseNet.skeleton(pose.keypoints));
```
Is this something we can package as a feature of the library? 

```javascript
function gotPoses(poses) {
 skeletons = poseNet.skeleton(poses);
}
```

or

```javascript
function gotPoses(poses, skeletons) {
 // callback passes arrays of both?
 // Does the end user always want the "skeleton" format?
}
```

At the very least we might want to add some comments / explanation of what's happening here.

cvalenzuela replied:
totally! I think the later is better, will update then

shiffman replied:
I'm a little confused as to the difference between a `pose` and a `skeleton` but I can take a look at the example and play with it a bit more after we merge!

cvalenzuela replied:
`pose` is an array of keypoints describing the body positions detected. 
```javascript
[ score: 0.9952507019042969, part: "nose", position: { x: 174.63247075574154, y: 226.84211836190062 } , ...]
```

`skeleton` is an array containing arrays of two elements describing connections between two keypoints
```javascript
[[part: "leftShoulder"], [ part: "rightShoulder" ], ... ]
```

shiffman replied:
Ah, that makes sense! Let me know when you are ready for me to merge!

cvalenzuela commented:
@shiffman, I just updated some new changes. I also added a couple of vanilla js examples.
I think we can merge this and so we can also merge https://github.com/ml5js/ml5-examples/pull/22

shiffman commented:
Yay!
